File: fg-events/fg-smn-event/src/fg_smn_event/smnevent.py
"""SMN event classes for FunctionGraph."""


class SMNEvent:
    """Represents a Simple Message Notification event for FunctionGraph."""

    def __init__(self, event):
        self._event = event or {}
        self._records = [SMNRecord(record) for record in self._event.get("record") or []]

    def get_records(self):
        return self._records

    def get_function_name(self):
        return self._event.get("functionname") or ""

    def get_request_id(self):
        return self._event.get("requestId") or ""

    def get_timestamp(self):
        return self._event.get("timestamp") or ""

    def to_json(self):
        return self._event


class SMNRecord:
    """Represents a single SMN event record."""

    def __init__(self, record):
        self._record = record or {}

    def get_event_subscription_urn(self):
        return self._record.get("event_subscription_urn") or ""

    def get_event_source(self):
        return self._record.get("event_source") or ""

    def get_smn_body(self):
        return SMNBody(self._record.get("smn"))

    def to_json(self):
        return self._record


class SMNBody:
    """Represents the SMN message body in a record."""

    def __init__(self, smn):
        self._smn = smn or {}

    def get_topic_urn(self):
        return self._smn.get("topic_urn") or ""

    def get_timestamp(self):
        return self._smn.get("timestamp") or ""

    def get_message_attributes(self):
        return self._smn.get("message_attributes")

    def get_message(self):
        return self._smn.get("message") or ""

    def get_type(self):
        return self._smn.get("type") or ""

    def get_message_id(self):
        return self._smn.get("message_id") or ""

    def get_subject(self):
        return self._smn.get("subject") or ""

    def to_json(self):
        return self._smn


__all__ = ["SMNEvent", "SMNRecord", "SMNBody"]

File: fg-events/fg-smn-event/src/smnevent.js
"use strict";

/**
 * @typedef {Object} SMNBodyJSON
 * @property {string} [topic_urn]
 * @property {string} [timestamp]
 * @property {Object | null} [message_attributes]
 * @property {string} [message]
 * @property {string} [type]
 * @property {string} [message_id]
 * @property {string} [subject]
 */

/**
 * @typedef {Object} SMNRecordJSON
 * @property {string} [event_subscription_urn]
 * @property {string} [event_source]
 * @property {SMNBodyJSON} [smn]
 */

/**
 * @typedef {Object} SMNEventJSON
 * @property {SMNRecordJSON[]} [record]
 * @property {string} [functionname]
 * @property {string} [requestId]
 * @property {string} [timestamp]
 */

/**
 * SMN Event Class
 * Represents a Simple Message Notification event for FunctionGraph
 */
class SMNEvent {
  /**
   * @param {SMNEventJSON} event
   */
  constructor(event) {
    this._event = event || {};

    this._records = [];
    for (const record of this._event.record || []) {
      this._records.push(new SMNRecord(record));
    }
  }

  /**
   * Returns the event records.
   * @returns {SMNRecord[]} Array of event records
   */
  getRecords() {
    return this._records;
  }

  /**
   * Returns the function name.
   * @returns {string} Function name
   */
  getFunctionName() {
    return this._event.functionname || "";
  }

  /**
   * Returns the request ID.
   * @returns {string} Request ID
   */
  getRequestId() {
    return this._event.requestId || "";
  }

  /**
   * Returns the timestamp.
   * @returns {string} Event timestamp
   */
  getTimestamp() {
    return this._event.timestamp || "";
  }

  /**
   * Converts the wrapped payload back to a plain JSON object.
   * @returns {SMNEventJSON} Payload as JSON object
   */
  toJSON() {
    return this._event;
  }

}

class SMNRecord {
  /**
   * @param {SMNRecordJSON} record
   */
  constructor(record) {
    this._record = record || {};
  }

  /**
   * Returns the event subscription URN.
   * @returns {string} Event subscription URN
   */
  getEventSubscriptionUrn() {
    return this._record.event_subscription_urn || "";
  }

  /**
   * Returns the event source.
   * @returns {string} Event source
   */
  getEventSource() {
    return this._record.event_source || "";
  }

  /**
   * Returns the SMN details from the record.
   * @returns {SMNBody} SMN details object
   */
  getSMNBody() {
    return new SMNBody(this._record.smn);
  }

  /**
   * Converts the wrapped payload back to a plain JSON object.
   * @returns {SMNRecordJSON} Payload as JSON object
   */
  toJSON() {
    return this._record;
  }
}

class SMNBody {
  /**
   * @param {SMNBodyJSON} smn
   */
  constructor(smn) {
    this._smn = smn || {};
  }

  /**
   * Returns the SMN topic URN.
   * @returns {string} Topic URN
   */
  getTopicUrn() {
    return this._smn.topic_urn || "";
  }

  /**
   * Returns the SMN timestamp.
   * @returns {string} SMN timestamp
   */
  getTimestamp() {
    return this._smn.timestamp || "";
  }

  /**
   * Returns the SMN message attributes.
   * @returns {Object|null} Message attributes
   */
  getMessageAttributes() {
    return this._smn.message_attributes;
  }

  /**
   * Returns the SMN message content.
   * @returns {string} Message content
   */
  getMessage() {
    return this._smn.message || "";
  }

  /**
   * Returns the SMN type.
   * @returns {string} SMN type
   */
  getType() {
    return this._smn.type || "";
  }

  /**
   * Returns the SMN message ID.
   * @returns {string} Message ID
   */
  getMessageId() {
    return this._smn.message_id || "";
  }

  /**
   * Returns the SMN message subject.
   * @returns {string} Message subject
   */
  getSubject() {
    return this._smn.subject || "";
  }

  /**
   * Converts the wrapped payload back to a plain JSON object.
   * @returns {SMNBodyJSON} Payload as JSON object
   */
  toJSON() {
    return this._smn;
  }
}

module.exports = { SMNEvent, SMNRecord, SMNBody };
